### PR TITLE
Adding statistical_analysis_of_llvm-ir_compilation blog post (GSoC 2024)

### PR DESCRIPTION
**Pull request by andrewkallai**
@asl here is my blog post for my gsoc project. I have tested with Hugo. Please let me know if you have feedback or changes are required. Thank you

### CODE REVIEW

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+Having a large IR dataset such as ComPile (linked below) allows for testing the LLVM compilation pipeline on a varied sample of IR. The size of this sample is sufficient to determine outlying IR modules. By identifying and examining such files using utilities which are being added to the LLVM IR Dataset Utils Repo (linked below), the causes of unexpected compilation times can be determined. Developers can then modify and improve the compilation pipeline accordingly.
+
+# Summary of Work
+
+The utilities added in PR37 (linked below) are intended to write each IR module to a tar file corresponding to a programming language. This is designed to be a multithreaded process that is agnostic to the number of files per language or the type of language. Executing the scripts correctly results in a file containing index values. Each file written to the tar files is indexed by its location in the HF dataset. The first file index begins at 1, so for N files, there are IR files from file1.bc to fileN.bc.
```
Comment (asl) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:18]: You can probably insert direct link here as well. And in few other places where you're referencing a PR.

**Comment by asl** [comment]
@boomanaiden154 Will you please review content-wise?

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
+---
+author: "Andrew Kallai"
+date: "2024-08-29"
+tags: ["GSoC", "llvm-ir-dataset-utils", "ComPile", "analysis"]
+title: "GSoC 2024: Statistical Analysis of LLVM-IR Compilation"
+---
+
+Welcome! My name is Andrew and I contributed to LLVM through the 2024 Google Summer of Code Program. My project is called [Statistical Analysis of LLVM-IR Compilation](https://summerofcode.withgoogle.com/programs/2024/projects/hquDyVBK). The objective of this project is to provide an analysis of the causes of unexpected compilation times.
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:8]: I would maybe reword the last couple words. Unexpected compilation times is kind of vague. Maybe something to the tune of "an analysis of where time spent in the optimization pipeline drastically increases in some instances and why this occurs".

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+# Background
+
+In principle, an LLVM IR bitcode file, or module, contains IR features that determine the behavior of the compiler optimization pipeline. By varying these features, the optimization pipeline, opt, can perform better or worse. More specifically, optimizations succeed in less or more time; the user can wait for a microsecond or a few minutes. LLVM compiler developers constantly edit the pipeline, so the performance of these optimizations can vary by time (sometimes by several seconds).
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:12]: I would clarify what you mean by opt performing better or worse. Specifically mention compilation time there, and then the sentence afterwards becomes reundant.

In the last sentence, I would say vary by version. Also, I wouldn't exactly quantify the time variance. It could be a lot more than a couple seconds in some instances. Something like "sometimes significantly" would cover most of your bases there I would think.

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+Having a large IR dataset such as [ComPile](https://huggingface.co/datasets/llvm-ml/ComPile) allows for testing the LLVM compilation pipeline on a varied sample of IR. The size of this sample is sufficient to determine outlying IR modules. By identifying and examining such files using utilities which are being added to the [LLVM IR Dataset Utils Repo](https://github.com/llvm-ml/llvm-ir-dataset-utils), the causes of unexpected compilation times can be determined. Developers can then modify and improve the compilation pipeline accordingly.
+
+# Summary of Work
+
+The utilities added in [PR37](https://github.com/llvm-ml/llvm-ir-dataset-utils/pull/37) are intended to write each IR module to a tar file corresponding to a programming language. This is designed to be a multithreaded process that is agnostic to the number of files per language or the type of language. Executing the scripts correctly results in a file containing index values. Each file written to the tar files is indexed by its location in the HF dataset. The first file index begins at 1, so for N files, there are IR files from file1.bc to fileN.bc.
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:18]: You can probably omit some of the details here. The last three sentences are probably all too much detail. Probably include some context on why moving things to local tar files was found to be useful.

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+# Summary of Work
+
+The utilities added in [PR37](https://github.com/llvm-ml/llvm-ir-dataset-utils/pull/37) are intended to write each IR module to a tar file corresponding to a programming language. This is designed to be a multithreaded process that is agnostic to the number of files per language or the type of language. Executing the scripts correctly results in a file containing index values. Each file written to the tar files is indexed by its location in the HF dataset. The first file index begins at 1, so for N files, there are IR files from file1.bc to fileN.bc. 
+
+The Makefile from [PR36](https://github.com/llvm-ml/llvm-ir-dataset-utils/pull/36) is responsible for carrying out the data collection. This data includes text segment size, user CPU instruction counts during compile time, IR feature counts sourced from LLVM function_analysis_properties, and maximum relative time pass names and percentage counts. The data can be extracted in parallel or serially and is stored in a CSV file.
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:20]: Can you put things like `func-properties` in backticks to mark them as a code block? Also, use the LLVM pass name here rather than the downstream python wrapper function.

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+Below is a modified version of code taken from the Makefile, which specifically illustrates what and how data is generated from the IR files:
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:22]: You can probably just get rid of this sentence and the Makefile. That's too much detail for a blog post.

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+   0.0014 (  0.1%)   0.0005 ( 13.3%)   0.0019 (  0.1%)   0.0020 (  0.1%)  EntryExitInstrumenterPass
+   0.0003 (  0.0%)   0.0000 (  0.0%)   0.0003 (  0.0%)   0.0003 (  0.0%)  CoroConditionalWrapper
+   2.2507 (100.0%)   0.0039 (100.0%)   2.2547 (100.0%)   2.2552 (100.0%)  Total
+```
+
+A user can visually see the distribution of these passes by using a profiling tool for .json files. The .json file for a given bitcode file is obtained by `clang -c -ftime-trace <file>`.
+
+The visualization of this output can be filtered to the passes of interest as in the following image:
+
+{{< figure src="/img/file4504_pass_trace.png" alt="ftime-trace of C IR Outlier" >}}
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:65]: Can you add some commentary on why this specific pass is interesting?

**Review by boomanaiden154 — COMMENTED** [review]
Couple smallish comments. Mostly looking good.

**Comment by asl** [comment]
@andrewkallai Any update?

**Comment by andrewkallai** [comment]
Added in the changes.

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+Table of outliers for C IR files:
+
+{{< figure src="/img/c_outliers.png" alt="C IR Outlier Files" >}}
+
+# Future Work
+
+It was discussed in PR 37 to consolidate the tar file creation into the dataset file writer Python script. This is a feature I wish to implement in order to speed up the tar file creation process by having the bitcode files written from memory to the tar instead of from memory, to disk, to tar.
+
+As mentioned, Julia IR was not analyzed. Modifying the scripts to include Julia IR results is desirable to make complete use of the dataset.
+Adding additional documentation for demonstration-of-use purposes could help clarify ways to use the tools.
```
Comment (boomanaiden154) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:65]: Maybe also mention down here planned analysis of the outliers?

**Review by boomanaiden154 — APPROVED** [review]
One nit, otherwise LGTM at this point as a broad overview of the project.

**Comment by andrewkallai** [comment]
Added the details.

**Review by boomanaiden154 — APPROVED** [review]

File: content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
+---
+author: "Andrew Kallai"
+date: "2024-08-29"
```
Comment (asl) [inline content/posts/2024-08-29-GSoC-2024-statistical_analysis_of_llvm-ir_compilation.md:3]:
```suggestion
date: "2024-09-23"
```

**Review by asl — APPROVED** [review]
Thanks! Scheduled for next monday (09/23)
